rfcs/2025-03-16-smp-queues.md

```diff
@@ -0,0 +1,163 @@
+
+-- with LinkId as entity ID, if queue does not allow sndSecure the command will fail, same as SKEY
+LGET :: Maybe SndPublicAuthKey -> Command Sender
+
+LINK :: QueueLink -> BrokerMsg
+```
+
+## Algorithm to prepare and to interpret queue link data.
+
+This follows the design proposed in [Short links](./2024-06-21-short-links.md) RFC.
+
+**Prepare queue link data**
+
+- the data blob owner generates X25519 key pair: `(k, pk)`.
+- private key `pk` will be included in the short link shared with the other party (only base64url encoded key bytes, not X509 encoding).
+- `HKDF(pk)` will be used to encrypt the link data with secret_box before storing it on the server.
+- the hash of public key `sha256(k)` will be used as ID by the owner to update the link data and by the accepting party to get link (`LSET` and `LGET` commands).
+
+**Retrieving queue link data**
+
+- the sender uses the public key `k` derived from the private key `pk` included in the link as entity ID to retrieve link data (the server will compute the ID used by the owner as `sha256(k)` and will be able to look it up). This provides the quality that the traffic of the parties has no shared IDs inside TLS. It also means that unlike message queue creation, the ID to retrieve the link data was never sent to the creator, and also is not known to the server in advance (the second part is only an observation, in itself it does not increase security, as server has access to an encrypted blob anyway).
+- note that the sender does not authorize the request to retrieve the blob, as it would not increase security unless a different key is used to authorize, and adding a key would increase link size.
+- server session keys with the sender will be `(sk, spk)`, where `sk` is public key shared with the sender during session handshake, and `spk` is the private key known only to the server.
+- this public key `k` will also be combined with server session key `spk` using `dh(k, spk)` to encrypt the response, so that there is no ciphertext in common in sent and received traffic for these blobs. Correlation ID will be used as a nonce for this encryption.
+- having received the blob, the client can now decrypt it using secret_box with `HKDF(pk)`.
+
+Using the same key as ID for the request, and also to additionally encrypt the response allows to use a single key in the link, without increasing the link size.
+
+## Threat model
+
+See [Short links](./2024-06-21-short-links.md) RFC.
+
+## Syntax for short links
+
+The proposed syntax:
+
+```abnf
+shortConnectionRequest = %s"https://" smpServerHost "/" connReqType "#/" linkKey [ "?" param *( "&" param ) ]
+smpServerHost = <hostname> ; RFC1123, RFC5891
+connReqType = %s"i" / %s"invitation" / %s"c" / %s"contact"
+linkKey = <base64url encoded private key data>
+
+param = fingerprintParam / hostsParam
+fingerprintParam = "f=" fingerprint ; server fingerprint
+hostsParam = "h=" <hostname> *( "," <hostname> ) ; additional hostnames, e.g. onion
```

Comment:
onion address to not be comma separated from smpServerHost? do we allow onion-only addresses currently? (I think yes - then the only address will be in the end)

Reply:
if it's onion-only it will be in smpServerHost: https://abcd.onion/i#/abcd